[WARC] Backward compatible storage of HTTP/2 headers

sebastian-nagel opened the pull request:
This PR addresses #828:
- the Okhttp protocol stores HTTP and TLS versions (if HTTP headers are stored)
- the WARC record formats normalize HTTP request and response status lines
  - replace HTTP version strings other than "HTTP/1.0" and "HTTP/1.1" 
- HTTP and TLS/SSL protocol versions are stored in the WARC header "WARC-Protocol" following roughly /iipc/warc-specifications#42 (protcol versions are stored as received from the protocol implementation in a single comma-separated value (no multiple headers)

Further improvements of the WARC module:
- add unit test for WARCHdfsBolt
  - write WARC file with info, request and response record
  - read and verify WARC file using [jwarc](/iipc/jwarc)
- extend GzipHdfsBolt from HdfsBolt (instead of AbstractHdfsBolt) and remove overridden methods with identical implementation in HdfsBolt
- add missing license headers

sebastian-nagel committed:
Okhttp protocol: store HTTP and TLS versions if headers are stored

Signed-off-by: Sebastian Nagel <sebastian@commoncrawl.org>

sebastian-nagel committed:
[WARC] Backward compatible storage of HTTP/2 headers
- WARC record formats to normalize HTTP request and response status lines
  - replace HTTP version strings other than "HTTP/1.0" and "HTTP/1.1"
  - normalize whitespace

Signed-off-by: Sebastian Nagel <sebastian@commoncrawl.org>

sebastian-nagel committed:
[WARC] Backward compatible storage of HTTP/2 headers
- store HTTP and TLS/SSL protocol versions in WARC header
  "WARC-Protocol"
- update README of WARC module, add information about
  storing HTTP messages

Signed-off-by: Sebastian Nagel <sebastian@commoncrawl.org>

sebastian-nagel committed:
Extend GzipHdfsBolt from HdfsBolt (instead of AbstractHdfsBolt)
and remove overridden methods with identical implementation in HdfsBolt

Signed-off-by: Sebastian Nagel <sebastian@commoncrawl.org>

sebastian-nagel committed:
Add unit test for WARCHdfsBolt
- write WARC file with info, request and response record
- read and verify WARC file using Jwarc

Signed-off-by: Sebastian Nagel <sebastian@commoncrawl.org>

sebastian-nagel committed:
TestOutputCollector to log exceptions

Signed-off-by: Sebastian Nagel <sebastian@commoncrawl.org>

sebastian-nagel committed:
WARC module: add missing license headers

Signed-off-by: Sebastian Nagel <sebastian@commoncrawl.org>

jnioche commented:
thanks @sebastian-nagel, looks great. Did you test these changes in a crawl?

sebastian-nagel commented:
> thanks @sebastian-nagel, looks great. Did you test these changes in a crawl?

Not yet, but will do the next days. So far, I've verified the WARC file generated by the unit using various validators (`warcio check`, `jwarc.tools.WarcTool validate`, [hanzo/warcvalid](/internetarchive/warctools/blob/master/hanzo/warcvalid.py)).

jnioche commented:
> > thanks @sebastian-nagel, looks great. Did you test these changes in a crawl?
> 
> Not yet, but will do the next days. So far, I've verified the WARC file generated by the unit using various validators (`warcio check`, `jwarc.tools.WarcTool validate`, [hanzo/warcvalid](/internetarchive/warctools/blob/master/hanzo/warcvalid.py)).

thanks. Let us know how it goes. Your changes look fine but we'll wait until they have been tried for real before merging. 
Will make some tiny superficial comments in the meantime

jnioche commented on core/src/main/java/com/digitalpebble/stormcrawler/protocol/okhttp/HttpProtocol.java:
```diff
@@ -570,12 +571,20 @@ public Response intercept(Interceptor.Chain chain) throws IOException {
             byte[] encodedBytesRequest =
                     Base64.getEncoder().encode(requestverbatim.toString().getBytes());
 
+            StringBuilder protocols = new StringBuilder(response.protocol().toString());
```
mark as _final_? same for _handshake_?
Doing so makes things a bit more explicit, possibly a bit faster but more importantly can highlight unexpected reassignments.

jnioche commented on external/warc/README.md:
````diff
@@ -153,11 +153,12 @@ Writing complete and valid WARC requires that HTTP headers, IP address and captu
-  - "http/1.1"
-```
+A note on the recording of HTTP requests and responses with StormCrawler and the WARC module:
+- the WARC file format is derived from the HTTP message format ([RFC 2616](https://www.ietf.org/rfc/rfc2616.txt)) and the WARC format as well as WARC readers require that HTTP requests and responses are recorded as HTTP/1.1 or HTTP/1.0 do. Therefor, the WARC WARCHdfsBolt writes binary HTTP formats (eg. [HTTP/2](https://en.wikipedia.org/wiki/HTTP/2)) as if they were HTTP/1.1. There is no need to l ...
````
` do. Therefor, `

remove do. Add e to therefore

jnioche commented on external/warc/src/main/java/com/digitalpebble/stormcrawler/warc/GzipHdfsBolt.java:
```diff
@@ -15,30 +15,21 @@
 
 /** Unlike the standard HdfsBolt this one writes to a gzipped stream with per-record compression. */
-public class GzipHdfsBolt extends AbstractHdfsBolt {
+public class GzipHdfsBolt extends HdfsBolt {
```
There must have been a reason why we chose to extend  _AbstractHdfsBolt_ but can't remember what it could have been. Or maybe there wasn't? It has been a while...

jnioche commented on external/warc/src/main/java/com/digitalpebble/stormcrawler/warc/WARCRecordFormat.java:
```diff
@@ -166,6 +187,31 @@ public static String fixHttpHeaders(String headers, int contentLength) {
                 boolean valid = true;
                 if (start == 0) {
                     // status line (without colon)
+                    String statusLine = headers.substring(0, lineEnd);
```
final?

jnioche commented on external/warc/src/main/java/com/digitalpebble/stormcrawler/warc/WARCRecordFormat.java:
```diff
@@ -166,6 +187,31 @@ public static String fixHttpHeaders(String headers, int contentLength) {
                     // status line (without colon)
+                    String statusLine = headers.substring(0, lineEnd);
+                    if (!STATUS_LINE_PATTERN.matcher(statusLine).matches()) {
+                        String[] parts = WS_PATTERN.split(headers.substring(0, lineEnd), 3);
```
_final String[] parts =_

jnioche commented on external/warc/src/main/java/com/digitalpebble/stormcrawler/warc/WARCRecordFormat.java:
```diff
@@ -370,6 +416,10 @@ public byte[] format(Tuple tuple) {
             }
             buffer.append("WARC-Truncated").append(": ").append(truncated).append(CRLF);
         }
+        String protocolVersions = metadata.getFirstValue(ProtocolResponse.PROTOCOL_VERSIONS_KEY);
```
final ?

jnioche commented on external/warc/src/main/java/com/digitalpebble/stormcrawler/warc/WARCRequestRecordFormat.java:
```diff
@@ -33,18 +51,12 @@ public byte[] format(Tuple tuple) {
-            }
-            httpheaders = headersVerbatim.getBytes();
         }
+        byte[] httpheaders = fixHttpHeaders(headersVerbatim).getBytes();
```
final?

jnioche commented on external/warc/src/main/java/com/digitalpebble/stormcrawler/warc/WARCRequestRecordFormat.java:
```diff
@@ -104,4 +116,78 @@ public byte[] format(Tuple tuple) {
+
+    private static String fixHttpHeaders(String headers) {
+        int start = 0, lineEnd = 0, last = 0, trailingCrLf = 0;
+        StringBuilder replace = new StringBuilder();
```
final? 
call it _replacement_ instead?

jnioche reviewed (commented):
Minor changes. This is a great PR, thanks. Brilliant to have more tests

sebastian-nagel committed:
WarcRecordFormat: avoid extra computation of block digest for WARC resource records
(without HTTP header block and payload digest are equal)

Signed-off-by: Sebastian Nagel <sebastian@commoncrawl.org>

sebastian-nagel committed:
Address code review comments
- make local variables final where applicable
- fix typos

Signed-off-by: Sebastian Nagel <sebastian@commoncrawl.org>

sebastian-nagel committed:
[WARC] Ensure that WARC response records include WARC-IP-Address,
WARC-Protocol and WARC-Truncated headers if applicable (fixes #1014)
- WARCRecordFormat: always pass protocolMDprefix when requesting metadata
  added by protocol implementations
- add unit tests to verify
  - that headers are included
  - the format of WARC resource records

Signed-off-by: Sebastian Nagel <sebastian@commoncrawl.org>

sebastian-nagel commented:
Updated PR:
- addressed code review comments (thanks @jnioche!)
- improved calculation of digests for WARC resource records (block and payload digest are the same)
- tested on Storm 2.4.0: written WARC files verified successfully
- addressed #1014 discovered during testing

sebastian-nagel committed:
Address code review comments
- make local variables final where applicable
  (also in class WARCRequestRecordFormat)

Signed-off-by: Sebastian Nagel <sebastian@commoncrawl.org>

jnioche commented:
Thanks @sebastian-nagel
